### PR TITLE
DAOS-12491 vos: handle DTX conflict for conditional updating ilog

### PR DESCRIPTION
**Pull request by Nasf-Fan**
When update ilog for update or punch, we fetch related ilog entries firstly, that may hits many uncommitted DTX entries on non-leaders. If it is for non-conditional modification, then it will not trigger DTX refresh nor client side RPC retry. But if it is for conditional modification, then the upper layer caller will get -DER_INPROGRESS. Under such case, related VOS logic need to collect those 'prepared' DTX entries information for DTX refresh to avoid related operation being blocked for long time.

Signed-off-by: Fan Yong <fan.yong@intel.com>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/vos/ilog.h [hunk 1]
```diff
@@ -242,14 +242,15 @@ ilog_fetch_move(struct ilog_entries *dest, struct ilog_entries *src);
  *  \param	root[in]	Pointer to log root
  *  \param	cbs[in]		Incarnation log transaction log callbacks
  *  \param	intent[in]	The intent of the operation
+ *  \param	has_conf[in]	Whether for conditional operation or not
```
Comment (jolivier23) [inline src/vos/ilog.h:245]: s/has_conf/has_cond

**Comment by Nasf-Fan** [comment]
@jolivier23 @wangdi1 @gnailzenh , would you please to help review the patch that is P1 blocker. Thanks!

**Comment by Nasf-Fan** [comment]
@wangdi1 @jolivier23 @gnailzenh , would you please to help review the patch that is for P1 blocker. Thanks!

File: src/vos/ilog.c [hunk 2]
```diff
@@ -1080,13 +1080,19 @@ ilog_fetch_move(struct ilog_entries *dest, struct ilog_entries *src)
...
+ilog_status_refresh(struct ilog_context *lctx, uint32_t intent, bool has_cond,
 		    struct ilog_entries *entries)
 {
 	struct ilog_priv	*priv = ilog_ent2priv(entries);
 	struct ilog_entry	 entry;
 	int32_t			 status;
 	bool			 same_intent = (intent == priv->ip_intent);
+	bool			 retry;
+
+	if ((intent == DAOS_INTENT_UPDATE || intent == DAOS_INTENT_PUNCH) && !has_cond)
```
Comment (wangdi1) [inline src/vos/ilog.c:1092]: can condition to be INTENT as well? as DAOS_COND_INTENT_UPDATE?

**Review by wangdi1 — APPROVED** [review]

File: src/vos/ilog.c (hunk 2 above)
Comment (Nasf-Fan) [reply src/vos/ilog.c:1092]: Yes, that is not impossible, but it seems not simplify the patch. Currently, we handle "intent" and "condition" as two independent parameters, not only in this function, but also some other places. That is how they are transferred from object layer downcall APIs. If not change related APIs, then we need to re-combine them and dispatch them to related functions inside vos. If it is necessary, we can consider that in subsequent patch.

**Review by gnailzenh — APPROVED** [review]

**Pull request merged**